Converting benchmarks to JMH

marko-bekhta opened the pull request:
So that's a converted benchmark. I think you might have better names for the modules :)

I run it with (starting at the root):
```
mvn clean install
cd jmh-benchmarks
mvn clean package -Phv-6.0-snapshot
java -jar target/bv-benchmarks-6.0.6-SNAPSHOT.jar
```

marko-bekhta committed:
Converting benchmarks to JMH

gsmet commented on jmh-benchmarks/pom.xml:
```diff
@@ -0,0 +1,192 @@
+
+	<profiles>
+		<profile>
+			<id>bval</id>
```
Let's name it bval-1.1

gsmet commented on pom.xml:
```diff
@@ -123,61 +108,11 @@
             </dependencies>
         </profile>
         <profile>
-            <id>hv-5.4</id>
```
Here is the workflow we need IMHO:
- generate the beans once and for all, compatible with all the BV implementations
- be able to run the benchmarks with the same beans on different versions.

I'm not sure what you did is compliant with this workflow?

So maybe we should stick to the validation-api dependencies as 1.1 provided in bean-generator (so that we are sure we only use 1.1 features AND we don't embark the 1.1 jar in the JMH jar).

I'm also wondering if we should make the bean-generator module a bean-generator-bv-1.1 module so that we could create a bv-2.0 in the future to also benchmark the BV 2.0 features.

And we could imagine having 2 different profiles to bundle either benchmarks in the JMH jar.

So for now, we would have `mvn -Phv-6.0-stable` to run the 1.1 benchmarks with HV 6.0 and in the future having `mvn -Pbv-2.0,hv-6.0-stable` to run our hypothetic BV 2.0 benchmarks on HV 6.0.

WDYT?

gsmet reviewed (commented):
@marko-bekhta added some comments. Maybe I missed something but I think we have an issue with the workflow.

marko-bekhta committed:
Various improvements

- Renamed generator module
- cleaned up profiles
- moved generator module into optional profile (so beans can be kept)

marko-bekhta commented:
@gsmet I've slightly modified what was already in there. Here's how the benchmarks can be run now:
```bash
# need install goal here so that generated beans get to local
# m2 repo and can be re-used in further builds 
# `generate-beans` profile inculdes bean generation module and
# generates new beans
# I've added snapshot profile as well to save some time but
# it can be executed as the others. It is not needed to add it
# when generating beans
mvn clean install -Pgenerate-beans -Phv-6.0-snapshot
# package benchmarks for other versions:
mvn package -Phv-6.0-stable
mvn package -Phv-5.4-stable

# need to go to benchmark directory as otherwise there will 
# be no target/ directory in the project root and benchmark 
# jar will fail to start with something like:
# Exception in thread "main" org.openjdk.jmh.runner.RunnerException: Can not touch the result file: target/jmh-results.json
cd jmh-benchmarks

# run benchmarks for different version one by one
java -jar target/bv-benchmarks-5.4.2.Final.jar
java -jar target/bv-benchmarks-6.0.5.Final.jar
java -jar target/bv-benchmarks-6.0.6-SNAPSHOT.jar
```

running this will create next benchmark jars in `jmh-benchmarks/target`:
```
-rw-rw-r-- 1 marko marko 5031064 Dec 13 14:31 bv-benchmarks-5.4.2.Final.jar
-rw-rw-r-- 1 marko marko 5400015 Dec 13 14:30 bv-benchmarks-6.0.5.Final.jar
-rw-rw-r-- 1 marko marko 5410974 Dec 13 14:30 bv-benchmarks-6.0.6-SNAPSHOT.jar
```

And just as a bonus :) here are some results from such run:
```
5.4.2.Final
# Run complete. Total time: 00:03:23

Benchmark                                            Mode  Cnt   Score   Error  Units
ParsingBeansSpeedBenchmark.testCascadedValidation   thrpt   20   6.337 ± 0.814  ops/s
RawValidationSpeedBenchmark.testCascadedValidation  thrpt   30  52.415 ± 4.990  ops/s


6.0.5.Final
# Run complete. Total time: 00:01:58

Benchmark                                            Mode  Cnt    Score   Error  Units
ParsingBeansSpeedBenchmark.testCascadedValidation   thrpt   20    8.923 ± 0.918  ops/s
RawValidationSpeedBenchmark.testCascadedValidation  thrpt   30  229.703 ± 3.417  ops/s


6.0.6-SNAPSHOT
# Run complete. Total time: 00:01:51

Benchmark                                            Mode  Cnt    Score   Error  Units
ParsingBeansSpeedBenchmark.testCascadedValidation   thrpt   20    8.870 ± 1.077  ops/s
RawValidationSpeedBenchmark.testCascadedValidation  thrpt   30  295.102 ± 4.857  ops/s
```

gsmet commented:
@marko-bekhta nice work.

Just wondering if it really makes sense to have a parent for the 2 modules.

Wouldn't it be more natural to do:
```
pushd bean-generator-bv-1.1
mvn clean install
popd
pushd jmh-benchmarks
mvn clean package -Phv-6.0-snapshot
mvn package -Phv-6.0-stable
mvn package -Phv-5.4-stable

# do your benchmarks here
```

WDYT?

marko-bekhta commented:
@gsmet hmmm, you're probably right... I cannot see any benefit of having these two under same parent. Maybe just some plugin version properties that are shared but, otherwise ... I'll remove the parent pom and cleanup it a bit more later today.

gsmet commented:
@marko-bekhta mmmh, good point for the properties.

gsmet commented:
@marko-bekhta Please check if there is not too much duplication between the 2 pom.xml if we separate them entirely. Might not be worth it after all.

gsmet commented on jmh-benchmarks/pom.xml:
```diff
@@ -0,0 +1,197 @@
+
+	<profiles>
+		<profile>
+			<id>bval1.1</id>
```
A dash here would be nice to be consistent with HV.

gsmet commented on bean-generator-bv-1.1/src/main/resources/generator.default.properties:
```diff
@@ -21,7 +21,7 @@
 #
 generator.rnd.seed=12
-generator.outputdir=./target/generated-sources/beangraph
+generator.outputdir=./bean-generator-bv-1.1/target/generated-sources/beangraph
```
So basically, this is the thing we have to change if we want to remove the parent module?

gsmet commented on jmh-benchmarks/pom.xml:
```diff
@@ -0,0 +1,197 @@
+									<!--
+                                        Shading signed JARs will fail without this.
+                                        http://stackoverflow.com/questions/999489/invalid-signature-file-when-attempting-to-run-a-jar
+                                    -->
```
The indentation is weird here.

marko-bekhta committed:
Made modules independent

- Removed the parent POM
- Pushed dependencies and plugins to corresponding modules
- Cleaned up remaining pom files
- Removed unnecessary sh scripts
- Updated Readme with new instructions on how to run benchmarks
- Updated ParsingBeansSpeedBenchmark
- Updated stable and snapshot versions for HV 6.0

marko-bekhta commented:
Finally had some time to spent on this.
@gsmet I thought about it and I think that not having the parent pom would be better. Mainly `generator.outputdir` convinced me to do so :smiley:. I've also done a few things here and there. All of them are mentioned in the commit message.

gsmet commented:
@marko-bekhta I added 2 more commits and merged. Thanks.

marko-bekhta commented:
@gsmet great, thanks!

gsmet commented:
@marko-bekhta FWIW I'm still not sure it's such a good idea to have the full version instead of what we have in HV (hv-5.4 for instance). I think having more consistency between the two would be great (and I must admit I prefer the HV way with shorter names). Didn't make the change as I thought you might have had good reasons to make it different?

marko-bekhta commented:
hmmm, thinking about it ... I could not remember any. I think that I just didn't add a property similar to what we have in HV:
```xml
<activation>
    <property>
        <name>validator</name>
        <value>hv-5.4</value>
    </property>
</activation>
```
to simply have less lines in pom.xml.
But I think you're right. It'll be better to keep this benchmark closer to what's already in HV. I can do the change + see what can be done about BV 2.0 model generation sometime this or next week (it's still "holiday season" for me 😃 and my schedule is not stable).

gsmet commented:
I think being consistent has some value. Moreover as we will probably do unit perf testing + BV benchmarks at the same time pretty often.

Not sure working on the BV 2.0 model has that much value for now. I would really like us to get this puppy (https://github.com/hibernate/hibernate-validator/pull/889) home while it's still relatively fresh. It's going to be harder and harder to get back to it and get it right :).
